### ISSUE TITLE
Explore Web component based primitives

### ISSUE THREAD

**Issue by vsjolander** [sebgroup/green#780]
Some functionality can and should be shared between React and Angular components. One idea is to build such components/functionality as Web components and this needs to be investigated further.

**Comment by patrikfogelstrand** [comment]
Has there been any more discussions in this area?

**Comment by patrikfogelstrand** [comment]
Regarding choice of web component helper, I think discussions around Lit vs. Stencil vs. another-web-component-helper can go on forever. 

We've had a great experience using Stencil lately, it is a really great way to create components, however it does have its drawbacks (as @splashdust mentioned). Primarily with long-lived issues (due to not enough resources working on it?) and that Stencil decided in the beginning decided to include Jest + Puppeteer inside of Stencil, meaning testing can sometimes (very rarely from what I've seen though) be a bit buggy, as the Stencil team can't keep up with including patches from Jest+Puppeteer. But having it included also means everything just works out of the box. Integration with Storybook is also great. 

Saying that Lit has bigger/better community than Stencil isn't really true, as Stencil has more than twice the amount of users than Lit does ([67k](https://github.com/ionic-team/stencil/network/dependents) vs [29.7k](https://github.com/lit/lit/network/dependents). Also, yes, those numbers don't say everything. 😄

**Comment by splashdust** [comment]
Great to know!

In large part, I based the decision to start looking at Lit on a [blog post](https://www.abeautifulsite.net/posts/moving-from-stencil-to-lit-element/) written by the author of Shoelace, a web component based UI framework. He was also a contributor of Stencil, and originally built Shoelace on Stencil, but later migrated to Lit and explains why in the blog post.
But I realize that the post is a bit dated now, and it looks like Ionic have been able to hire some new talent since then also, so perhaps they've been able to get past some of those issues by now.

So far we're only dipping our toes in this, and trying to figure out which potential obstacles we'll run into with web components going forward. Once we've built a few complete components in Green and have a clearer view of the full landscape, so to speak, we can make a final decision on which compiler/library fits best for Green!

**Comment by patrikfogelstrand** [comment]
Yeah, Stencil has lately been getting a lot more attention, especially with v. 3.x.x being released. Testing is still quite uncertain, as they still think about [decoupling Stencil from the test frameworks](https://github.com/ionic-team/stencil/issues/3156).

**Comment by patrikfogelstrand** [comment]
After our discussions last week @splashdust @astrit, should we perhaps close this issue? Base atom components in Lit sounds like a good plan to me 👍

**Comment by splashdust** [comment]
Agreed!

To summarize:

The new architectural vision for Green, going forward, will be to build all underlying functionality as native Web Components, utilising Shadow DOM for style and markup encapsulation. The components will be built using Lit, in order to reduce the amount of boiler plate, and to avoid re-inventing complex rendering logic, property binding, and so on.

The shape of the architecture will look something like this:

```mermaid
flowchart TD
    DT[Design Tokens]---AP
    M[Markup]---AP
    S[Styles]---AP
    AP[Atomic Primitives]---|Composition|CP[Compound Primitives]---|Public API Boundry|WC[Web Components]
    WC---A[Angular]
    WC---R[React]
```

There will be three different options for consuming components: As native Web Components or through React or through Angular. The React and Angular components will essentially just be thin wrappers over the native ones, possibly with a translation layer for backwards compatibility.

The Web Components themselves will be composed together out of smaller, more atomic components (primitives) which will also be implemented as Web Components. But unlike the full components, these primitives will not be considered part of the public API.
